### PR TITLE
Firewall config enhancement

### PR DESCRIPTION
**Pull request by pierantoniomerlino**
This PR tries to improve the robustness of the firewall management in Kura. Since on a device other processes (as Docker) can change the firewall rules, Kura has to be aware of that or simply ignore the changes.

**Related Issue:** This PR fixes/closes N/A

**Description of the solution adopted:** The firewall configuration is now added to specific custom `iptables` chains:

- **input-kura** for the INPUT rules
- **output-kura** for the OUTPUT rules
- **forward-kura** for the FORWARD rules
- **prerouting-kura** for the PREROUTING nat rules
- **postrouting-kura** for the POSTROUTING nat rules

The `firewall.service` is updated to check at boot if the new chains are present and create them if needed.
The `LinuxFirewall` and `IptablesConfig` classes are modified accordingly with these new chains. Moreover, Kura now saves the firewall config before applying the new rules, to keep the ones added by other processes.

Finally, the `docker0` interface added by Docker is configured as `NotManaged` by Kura.

### CODE REVIEW

**Comment by MMaiero** [comment]
@mstankovic Can you please start having a look at it?

**Comment by mstankovic** [comment]
I checked this one.

verifies that rules added by kura satisfied changes:

```
pi@raspberrypi:~ $ sudo iptables -S
-P INPUT DROP
-P FORWARD DROP
-P OUTPUT ACCEPT
-N input-kura
-N forward-kura
-N output-kura
-A INPUT -j input-kura
-A FORWARD -j forward-kura
-A OUTPUT -j output-kura
-A input-kura -i wlan0 -p tcp -m tcp --dport 8000 -j ACCEPT
-A input-kura -i eth0 -p tcp -m tcp --dport 8000 -j ACCEPT
-A input-kura -i wlan0 -p udp -m udp --dport 67 -j ACCEPT
-A input-kura -i eth0 -p udp -m udp --dport 67 -j ACCEPT
-A input-kura -i wlan0 -p udp -m udp --dport 53 -j ACCEPT
-A input-kura -i eth0 -p udp -m udp --dport 53 -j ACCEPT
-A input-kura -s 127.0.0.1/32 -p tcp -m tcp --dport 5002 -j ACCEPT
-A input-kura -i wlan0 -p tcp -m tcp --dport 1450 -j ACCEPT
-A input-kura -i eth0 -p tcp -m tcp --dport 1450 -j ACCEPT
-A input-kura -s 10.234.0.0/16 -p tcp -m tcp --dport 443 -j ACCEPT
-A input-kura -i wlan0 -p tcp -m tcp --dport 443 -j ACCEPT
-A input-kura -i eth0 -p tcp -m tcp --dport 443 -j ACCEPT
-A input-kura -s 10.234.0.0/16 -p tcp -m tcp --dport 4443 -j ACCEPT
-A input-kura -i wlan0 -p tcp -m tcp --dport 4443 -j ACCEPT
-A input-kura -i eth0 -p tcp -m tcp --dport 4443 -j ACCEPT
-A input-kura -p tcp -m tcp --dport 22 -j ACCEPT
-A input-kura -i lo -j ACCEPT
-A input-kura -m state --state RELATED,ESTABLISHED -j ACCEPT
-A input-kura -p icmp -m icmp --icmp-type 8 -m state --state NEW,RELATED,ESTABLISHED -j ACCEPT
-A input-kura -j RETURN
-A forward-kura -i eth0 -o wlan0 -m state --state RELATED,ESTABLISHED -j ACCEPT
-A forward-kura -i wlan0 -o eth0 -j ACCEPT
-A forward-kura -j RETURN
-A output-kura -p icmp -m icmp --icmp-type 0 -m state --state RELATED,ESTABLISHED -j ACCEPT
-A output-kura -j RETURN
```

- also run all Firewall tests for Kura and all have passed!

- noticed that after Kura uninstallation still have Kura specific rules: 

```
Uninstalling KURA... Done!
pi@raspberrypi:~ $ sudo iptables -S
-P INPUT DROP
-P FORWARD DROP
-P OUTPUT ACCEPT
-N input-kura
-N forward-kura
-N output-kura
-A INPUT -j input-kura
-A FORWARD -j forward-kura
-A OUTPUT -j output-kura
-A input-kura -i lo -j ACCEPT
-A input-kura -m state --state RELATED,ESTABLISHED -j ACCEPT
-A input-kura -p icmp -m icmp --icmp-type 8 -m state --state NEW,RELATED,ESTABLISHED -j ACCEPT
-A input-kura -p tcp -m tcp --dport 22 -j ACCEPT
-A input-kura -i wlan0 -p udp -m udp --dport 53 -j ACCEPT
-A input-kura -i eth0 -p udp -m udp --dport 53 -j ACCEPT
-A input-kura -i wlan0 -p udp -m udp --dport 67 -j ACCEPT
-A input-kura -i eth0 -p udp -m udp --dport 67 -j ACCEPT
-A input-kura -s 10.234.0.0/16 -p tcp -m tcp --dport 443 -j ACCEPT
-A input-kura -i wlan0 -p tcp -m tcp --dport 443 -j ACCEPT
-A input-kura -p tcp -m tcp --dport 443 -j ACCEPT
-A input-kura -i wlan0 -p tcp -m tcp --dport 1450 -j ACCEPT
-A input-kura -i eth0 -p tcp -m tcp --dport 1450 -j ACCEPT
-A input-kura -s 10.234.0.0/16 -p tcp -m tcp --dport 4443 -j ACCEPT
-A input-kura -i wlan0 -p tcp -m tcp --dport 4443 -j ACCEPT
-A input-kura -i eth0 -p tcp -m tcp --dport 4443 -j ACCEPT
-A input-kura -s 127.0.0.1/32 -p tcp -m tcp --dport 5002 -j ACCEPT
-A input-kura -s 172.50.1.0/24 -p tcp -m tcp --dport 5400 -j ACCEPT
-A input-kura -i wlan0 -p tcp -m tcp --dport 8000 -j ACCEPT
-A input-kura -i eth0 -p tcp -m tcp --dport 8000 -j ACCEPT
-A input-kura -j RETURN
-A forward-kura -i wlan0 -o eth0 -j ACCEPT
-A forward-kura -i eth0 -o wlan0 -m state --state RELATED,ESTABLISHED -j ACCEPT
-A forward-kura -j RETURN
-A output-kura -p icmp -m icmp --icmp-type 0 -m state --state RELATED,ESTABLISHED -j ACCEPT
-A output-kura -j RETURN
```

let me know if you need something additional

**Comment by MMaiero** [comment]
@pierantoniomerlino How is the update managed here?
Do the code manage the transition or do we need a note on the fact that this is a completely new flow that is incompatible, for example with the old snapshots/configurations?

**Comment by pierantoniomerlino** [comment]
@MMaiero I think that the snapshot will not change after this PR, since it doesn't contains any reference to the firewall chains. They are managed internally by Kura.
Regarding the update, it'd be tested. However, I think that we need a custom update for that...
Finally, it's fine that the firewall rules are still there after the uninstallation, since Kura doesn't delete any configuration.

**Review by MMaiero — APPROVED** [review]

**Comment by MMaiero** [comment]
@pierantoniomerlino please review the rebase

File: kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/FilterForwardChainRuleTest.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package org.eclipse.kura.linux.net.iptables;
```
Comment (MMaiero) [inline kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/FilterForwardChainRuleTest.java:13]: Please add the copyright header

File: kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/FirewallTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package org.eclipse.kura.linux.net.iptables;
```
Comment (MMaiero) [inline kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/FirewallTestUtils.java:13]: Please review the copyright header placing

File: kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/LocalRuleTest.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package org.eclipse.kura.linux.net.iptables;
```
Comment (MMaiero) [inline kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/LocalRuleTest.java:13]: Please add the copyright header

File: kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/NatPostroutingChainRuleTest.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package org.eclipse.kura.linux.net.iptables;
```
Comment (MMaiero) [inline kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/NatPostroutingChainRuleTest.java:13]: Please add the copyright header

File: kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/NatPreroutingChainRuleTest.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package org.eclipse.kura.linux.net.iptables;
```
Comment (MMaiero) [inline kura/test/org.eclipse.kura.linux.net.test/src/test/java/org/eclipse/kura/linux/net/iptables/NatPreroutingChainRuleTest.java:13]: Please add the copyright header

**Pull request merged**